helion/autotuner/cache.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import annotations
+
+import dataclasses
+import functools
+import hashlib
+import inspect
+import logging
+import os
+from pathlib import Path
+import textwrap
+from typing import TYPE_CHECKING
+from typing import Sequence
+
+import torch
+
+from .._utils import counters
+from ..runtime.config import Config
+
+if TYPE_CHECKING:
+    from ..runtime.kernel import Kernel
+
+log: logging.Logger = logging.getLogger(__name__)
+
+"""
+TODO(oulgen)
+- Allow user defined cache keys that can be passed on @helion.kernel
+- Add import/export for set of configs
+"""
+
+
+@dataclasses.dataclass(frozen=True)
+class AutotuneCacheKey:
+    """
+    helion_key: Hash of source code of Helion
+    torch_key: Hash of source code of PyTorch
+    system_hash: Hash of system information,
+        including Triton, current device, cuda/rocm arch version
+    function_source_hash: Hash of source code of input Helion kernel
+    input_dtypes: dtypes of input tensors
+    input_shapes: shapes of input tensors
+    """
+
+    helion_key: str
+    torch_key: str
+    system_hash: str
+    kernel_source_hash: str
+    input_dtypes: list[tuple[int, torch.dtype]]
+    input_shapes: list[tuple[int, torch.Size]]
+
+    def stable_hash(self) -> str:
+        return hashlib.sha256(repr(self).encode("utf-8")).hexdigest()
+
+
+class AutotuneCache:
+    def __init__(self, kernel: Kernel, args: Sequence[object]) -> None:
+        self.key: AutotuneCacheKey = AutotuneCache._generate_key(kernel, args)
+
+    @staticmethod
+    def _generate_key(kernel: Kernel, args: Sequence[object]) -> AutotuneCacheKey:
+        from torch._inductor.codecache import CacheBase
+        from torch._inductor.codecache import torch_key
+
+        kernel_source = textwrap.dedent(inspect.getsource(kernel.fn))
+        kernel_source_hash = hashlib.sha256(kernel_source.encode("utf-8")).hexdigest()
+
+        input_dtypes = []
+        input_shapes = []
+
+        for idx, a in enumerate(args):
+            if isinstance(a, torch.Tensor):
+                input_dtypes.append((idx, a.dtype))
+                input_shapes.append((idx, a.shape))
```

Comment:
Maybe we should mimic/reuse some of the key generation infra in kernel.py (for the in-memory cache).  IMO the memory key should be a subset of the disk key.  I think right now the memory key includes some extra stuff.